src/core/struct_tree.js

```diff
@@ -67,6 +69,439 @@ class StructTreeRoot {
+    newRefs,
+    buffer
+  ) {
+    const objr = Name.get("OBJR");
+    let nextKey = -Infinity;
+
+    for (const [pageIndex, elements] of newAnnotationsByPage) {
+      const [, pageRef] = await catalog.getPageDict(pageIndex);
+      for (const {
+        accessibilityData: { type, title, lang, alt, expanded, actualText },
+        ref,
+        parentTreeId,
+        structTreeParent,
+      } of elements) {
+        nextKey = Math.max(nextKey, parentTreeId);
+
+        const tagRef = xref.getNewTemporaryRef();
+        const tagDict = new Dict(xref);
+
+        // The structure type is required.
+        tagDict.set("S", Name.get(type));
+
+        if (title) {
+          tagDict.set("T", title);
+        }
+        if (lang) {
+          tagDict.set("Lang", lang);
+        }
+        if (alt) {
+          tagDict.set("Alt", alt);
+        }
+        if (expanded) {
+          tagDict.set("E", expanded);
+        }
+        if (actualText) {
+          tagDict.set("ActualText", actualText);
+        }
+
+        if (structTreeParent) {
+          await this.#updateParentTag(
+            structTreeParent,
+            tagDict,
+            tagRef,
+            xref,
+            newRefs,
+            buffer
+          );
+        } else {
+          tagDict.set("P", structTreeRootRef);
+        }
+
+        const objDict = new Dict(xref);
+        tagDict.set("K", objDict);
+        objDict.set("Type", objr);
+        objDict.set("Pg", pageRef);
+        objDict.set("Obj", ref);
+
+        buffer.length = 0;
+        await writeObject(tagRef, tagDict, buffer, xref);
+        newRefs.push({ ref: tagRef, data: buffer.join("") });
+
+        nums.push(parentTreeId, tagRef);
+        kids.push(tagRef);
+      }
+    }
+    return nextKey + 1;
+  }
+
+  static #collectParents(elements, xref, pageDict, pageIndex, parentTree) {
+    const idToElement = new Map();
+    for (const element of elements) {
+      if (element.structTreeParentId) {
+        const id = parseInt(element.structTreeParentId.split("_mc")[1], 10);
+        idToElement.set(id, element);
+      }
+    }
+
+    const id = pageDict.get("StructParents");
+    const numberTree = new NumberTree(parentTree, xref);
+    const parentArray = numberTree.get(id);
+    if (Array.isArray(parentArray)) {
+      for (const kidRef of parentArray) {
+        if (!(kidRef instanceof Ref)) {
+          continue;
+        }
+        const pageKid = xref.fetch(kidRef);
+        const k = pageKid.get("K");
+        if (Number.isInteger(k)) {
+          const element = idToElement.get(k);
+          if (element) {
+            const parentRef = pageKid.getRaw("P");
+            const parentDict = xref.fetchIfRef(parentRef);
+            if (parentRef instanceof Ref && parentDict instanceof Dict) {
+              // It should always the case, but we check just in case.
+              element.structTreeParent = { ref: kidRef, dict: pageKid };
+            }
+          }
+          continue;
```

Comment:
Given that this code is effectively duplicated below, would it make sense to add a helper function above?
E.g. something like this:
```javascript
const updateElement = kid => {
  const element = idToElement.get(kid);
  if (element) {
    const parentRef = pageKid.getRaw("P");
    const parentDict = xref.fetchIfRef(parentRef);
    if (parentRef instanceof Ref && parentDict instanceof Dict) {
      // It should always the case, but we check just in case.
      element.structTreeParent = { ref: kidRef, dict: pageKid };
    }
    return true;
  }
  return false;
};
```
which would simplify this code to just:
```suggestion
          updateElement(k);
          continue;
```